superset-frontend/src/explore/components/controls/DndColumnSelectControl/DndFilterSelect.tsx

```diff
@@ -0,0 +1,334 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with work for additional information
+ * regarding copyright ownership.  The ASF licenses file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import React, { useEffect, useMemo, useState } from 'react';
+import { logging, SupersetClient } from '@superset-ui/core';
+import { ColumnMeta, Metric } from '@superset-ui/chart-controls';
+import { Tooltip } from 'src/common/components/Tooltip';
+import { OPERATORS } from 'src/explore/constants';
+import { OptionSortType } from 'src/explore/types';
+import { DndFilterSelectProps, FilterOptionValueType } from './types';
+import AdhocFilterPopoverTrigger from '../FilterControl/AdhocFilterPopoverTrigger';
+import OptionWrapper from './components/OptionWrapper';
+import DndSelectLabel from './DndSelectLabel';
+import AdhocFilter, {
+  CLAUSES,
+  EXPRESSION_TYPES,
+} from '../FilterControl/AdhocFilter';
+import AdhocMetric from '../MetricControl/AdhocMetric';
+import {
+  DatasourcePanelDndItem,
+  DndItemValue,
+} from '../../DatasourcePanel/types';
+import { DndItemType } from '../../DndItemType';
+
+const isDictionaryForAdhocFilter = (value: FilterOptionValueType) =>
+  value && !(value instanceof AdhocFilter) && value.expressionType;
```

Comment:
I assume checking for value here isn't necessary as `value` is always an object (=always truthy)?

Reply:
I removed `value` and used `value?.expressionType` instead, just in case it was null or undefined